app/code/Magento/Inventory/Model/SourceRepository.php

```diff
@@ -0,0 +1,190 @@
+        $this->carrierLinkCollectionFactory = $carrierLinkCollectionFactory;
+        $this->sourceSearchResultsFactory = $sourceSearchResultsFactory;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function save(SourceInterface $source)
+    {
+        try {
+            /** @var SourceInterface|AbstractModel $source */
+            $this->resourceSource->save($source);
+
+            /** @var SourceCarrierLinkInterface|AbstractModel $carrierLink */
+            foreach ($source->getCarrierLinks() as $carrierLink) {
+                $carrierLink->setData(SourceInterface::SOURCE_ID, $source->getSourceId());
+                $this->resourceSourceCarrierLink->save($carrierLink);
+            }
+
+            return $source->getSourceId();
+        } catch (\Exception $exception) {
+            $this->logger->error($exception->getMessage());
+            throw new CouldNotSaveException(__('Could not save source'));
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function get($sourceId)
+    {
+        /** @var SourceInterface|AbstractModel $source */
+        $source = $this->sourceFactory->create();
+        $this->resourceSource->load($source, $sourceId, SourceInterface::SOURCE_ID);
+        $this->addCarrierLinks($source);
+
+        if (!$source->getSourceId()) {
+            throw NoSuchEntityException::singleField(SourceInterface::SOURCE_ID, $sourceId);
+        }
+
+        return $source;
+    }
+
+    /**
+     * @param SourceInterface $source
+     */
+    private function addCarrierLinks(SourceInterface $source)
+    {
+        /** @var ResourceSourceCarrierLink\Collection $collection */
+        $collection = $this->carrierLinkCollectionFactory->create();
+
+        $searchCriteria = $this->searchCriteriaBuilder
+            ->addFilter(SourceInterface::SOURCE_ID, $source->getSourceId())
+            ->create();
+
+        $this->collectionProcessor->process($searchCriteria, $collection);
+        $source->setCarrierLinks($collection->getItems());
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getList(SearchCriteriaInterface $searchCriteria = null)
+    {
+        /** @var \Magento\Inventory\Model\Resource\Source\Collection $collection */
+        $collection = $this->collectionFactory->create();
+
+        /** @var SourceSearchResultsInterface $searchResults */
+        $searchResults = $this->sourceSearchResultsFactory->create();
+
+        // if there is a searchCriteria defined, use it to add its creterias to the collection
+        if (!is_null($searchCriteria)) {
+            $this->collectionProcessor->process($searchCriteria, $collection);
+            $searchResults->setSearchCriteria($searchCriteria);
+        }
+
+        $searchResults->setItems($collection->getItems());
+        $searchResults->setTotalCount($collection->getSize());
+        return $searchResults;
```

Comment:
looks like current implementation of getList method would not add SourceCarrierLinks to retrieved sources